<?xml version="1.0" encoding="UTF-8"?>

<launch>
  <arg name="generation" default="Classic"/>
  <arg name="simulated" default="True"/>
  <arg name="config" default="$(find teleop)/config/console-full-system-simulated.json"/>

  <!-- Patient Cart -->
  <!-- SUJs -->
  <arg name="_urdf" default="$(find dvrk_model)/urdf/$(arg generation)/SUJ.urdf.xacro"/>
  <group ns="SUJ">
    <arg name="rate" default="20"/>
    <param name="robot_description" command="rosrun xacro xacro $(arg _urdf)"/>
    <param name="rate" value="$(arg rate)"/>
    <rosparam param="source_list" subst_value="True">
      [/SUJ/PSM1/measured_js, /SUJ/PSM2/measured_js, /SUJ/PSM3/measured_js, /SUJ/ECM/measured_js]
    </rosparam>
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
  </group>

  <!-- Active Arms -->
  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="generation" value="$(arg generation)"/>
    <arg name="arm" value="PSM1"/>
    <arg name="urdf_file" value="psm"/>
    <arg name="suj" value="True"/>
  </include>

  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="generation" value="$(arg generation)"/>
    <arg name="arm" value="PSM2"/>
    <arg name="urdf_file" value="psm"/>
    <arg name="suj" value="True"/>
  </include>

  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="generation" value="$(arg generation)"/>
    <arg name="arm" value="PSM3"/>
    <arg name="urdf_file" value="psm"/>
    <arg name="suj" value="True"/>
  </include>

  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="generation" value="$(arg generation)"/>
    <arg name="arm" value="ECM"/>
    <arg name="urdf_file" value="ecm"/>
    <arg name="suj" value="True"/>
  </include>

  <!-- Surgeon Console -->
  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="arm" value="MTML"/>
    <arg name="generation" value="$(arg generation)"/>
  </include>

  <include file="$(find dvrk_model)/launch/arm_state_publishers.launch">
    <arg name="arm" value="MTMR"/>
    <arg name="generation" value="$(arg generation)"/>
  </include>

  <!-- RViz -->
  <node name="rviz" 
        pkg="rviz" 
        type="rviz"  
        args="-d $(find teleop)/rviz/full_system.rviz"/>

  <node name="dvrk_patient_cart_node"
        if="$(arg simulated)"
        pkg="dvrk_robot"
        type="dvrk_console_json"
        args="-j $(arg config)"
        output="screen"/>

  <!-- Static Transform Publisher -->
  <node name="tf_world_to_cart" 
        pkg="tf2_ros" 
        type="static_transform_publisher" 
        args="0 0 0.1678 0 0 0 world Cart" />
</launch>